Source/WebCore/platform/SimpleCaretAnimator.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2004-2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+ */
+
+#include "config.h"
+#include "SimpleCaretAnimator.h"
+
+#include "RenderTheme.h"
+
+namespace WebCore {
+
+SimpleCaretAnimator::SimpleCaretAnimator(CaretAnimationClient& client)
+    : CaretAnimator(client)
+{
+}
+
+SimpleCaretAnimator::~SimpleCaretAnimator() = default;
+
+void SimpleCaretAnimator::serviceAnimation(ReducedResolutionSeconds currentTime)
+{
+    if (isCaretBlinkingSuspended() && visible())
+        return;
+
+    const auto caretBlinkInterval = RenderTheme::singleton().caretBlinkInterval();
+
+    if (currentTime - m_lastTimeCaretPaintWasToggled >= caretBlinkInterval - (caretBlinkInterval * 0.01)) {
```

Comment:
It's because previously, `m_isVisible` was just toggled every time the timer was fired. However now, it uses the Page::updateRendering / scheduleRenderingUpdate pipeline, which means that the timing will not be exactly the interval of the timer, and the function may be called at any time by external events, so the delta time must now be consulted to determine if it should be toggled or not.

Reply:
Yes but the 1% here makes it clear we have racy code that won't do the right thing in all cases. We need to do this some way that doesn't require a fudge factor.

Reply:
Ah ok, I'll work on fixing this, thanks!

Reply:
I think we want to subtract by a more predictable quantum based on timer resolution. Not 1% of the interval, which works by accident.

Reply:
This might be timer slop due to alignment, or maybe it's the "reduced resolution" part of `ReducedResolutionSeconds`. If you wait at least 500ms do you always get an update after that?

Reply:
@darinadler @smfr It actually works now with  Page::updateRendering without the threshold factor, since now the timer is a one-shot timer instead of a repeating one!